include/alpaka/pltf/PltfGenericSycl.hpp

```diff
@@ -18,89 +18,164 @@
+#        include <iostream>
+#    endif
+#    include <mutex>
+#    include <optional>
 #    include <sstream>
 #    include <stdexcept>
 #    include <vector>
 
 namespace alpaka::experimental
 {
     //! The SYCL device manager.
-    class PltfGenericSycl : public concepts::Implements<ConceptPltf, PltfGenericSycl>
+    template<typename TSelector>
+    class PltfGenericSycl : public concepts::Implements<ConceptPltf, PltfGenericSycl<TSelector>>
     {
+        using RecursiveLock = std::unique_lock<std::recursive_mutex>;
+
     public:
         PltfGenericSycl() = delete;
+
+        [[nodiscard]] static auto syclPlatform() -> sycl::platform&
+        {
+            auto lock = RecursiveLock{mutex};
+
+            if(!initialized)
+                initialize();
+
+            return platform_opt.value();
+        }
+
+        [[nodiscard]] static auto syclDevices() -> std::vector<sycl::device>&
+        {
+            auto lock = RecursiveLock{mutex};
+
+            if(!initialized)
+                initialize();
+
+            return devices_opt.value();
+        }
+
+        [[nodiscard]] static auto syclContext() -> sycl::context&
+        {
+            auto lock = RecursiveLock{mutex};
+
+            if(!initialized)
+                initialize();
+
+            return context_opt.value();
+        }
+
+        static auto initialize() -> void
+        {
+            auto lock = RecursiveLock{mutex};
+
+            if(initialized)
+                return;
+
+            platform_opt = sycl::platform{TSelector{}};
+            devices_opt = platform_opt.value().get_devices();
+            context_opt = sycl::context{
+                devices_opt.value(),
+                [](sycl::exception_list exceptions)
+                {
+                    auto ss_err = std::stringstream{};
+                    ss_err << "Caught asynchronous SYCL exception(s):\n";
+                    for(std::exception_ptr e : exceptions)
+                    {
+                        try
+                        {
+                            std::rethrow_exception(e);
+                        }
+                        catch(sycl::exception const& err)
+                        {
+                            ss_err << err.what() << " (" << err.code() << ")\n";
+                        }
+                    }
+                    throw std::runtime_error(ss_err.str());
+                }};
+
+            initialized = true;
+        }
+
+        static auto reset() -> void
+        {
+            auto lock = RecursiveLock{mutex};
+
+            if(!initialized)
+                return;
+
+            context_opt.reset();
+            devices_opt.reset();
+            platform_opt.reset();
+            initialized = false;
+        }
+
+    private:
+#    if BOOST_COMP_CLANG
+#        pragma clang diagnostic push
+#        pragma clang diagnostic ignored "-Wexit-time-destructors"
+#    endif
+        inline static std::recursive_mutex mutex;
```

Comment:
I don't see why a recursive mutex is needed here. Why can't this just be an ordinary `std::mutex`? Where is the reentrancy here?

Reply:
I've chosen `recursive_mutex` because both `initialize()` and the entire `syclObject()` methods need to be guarded. Otherwise, T1 might call `reset()` *after* T2 is past the `initialize()` call.

What I would actually need here is a `std::recursive_shared_mutex` (for shared access in `syclObject()` and exclusive access in `initialize()` but sadly the standard library doesn't provide this.